### ISSUE TITLE
Inconsistency between FQN and aliased imports

### ISSUE THREAD

**Issue by JamyDev** [sourcegraph/scip-go#34]
If you do an aliased import, we get a local definition:
```
   pb "gogoproto/some/package/some-name-proto"
// ^^ definition local 0
//     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ reference scip-go gomod gogoproto/some/package/some-name-proto 0.0.1 gogoproto/some/package/some-name-proto/
```

When importing a package as a FQN we don't get a definition for the actual variable that is used inside the code:
```
  import (
   "context"
//  ^^^^^^^ reference scip-go gomod github.com/golang/go/src 1.20.4 context/
   "time"
//  ^^^^ reference scip-go gomod github.com/golang/go/src 1.20.4 time/
   "golang.org/x/mod/modfile"
//  ^^^^^^^^^^^^^^^^^^^^^^^^ reference scip-go gomod golang.org/x/mod/modfile 0.0.1 golang.org/x/mod/modfile/
  )
```

Expected:

```
  import (
   "golang.org/x/mod/modfile"
//  ^^^^^^^^^^^^^^^^^^^^^^^^ reference scip-go gomod golang.org/x/mod/modfile 0.0.1 golang.org/x/mod/modfile/
//                   ^^^^^^^ reference ...
  )
```

I don't know what the best way of defining this is, but it may even make sense to split up the entire import path into its constituent parts like so. This would match how java does packages in scip-java.
```
  import (
   "golang.org/x/mod/modfile"
//  ^^^^^^^^^^^              reference scip-go gomod golang.org 0.0.1 golang.org/
//             ^^            reference scip-go gomod golang.org/x 0.0.1 golang.org/x/
//               ^^^^        reference scip-go gomod golang.org/x/mod 0.0.1 golang.org/x/mod/
//                   ^^^^^^^ reference scip-go gomod golang.org/x/mod/modfile 0.0.1 golang.org/x/mod/modfile/
//                   ^^^^^^^ definition local 0
  )
```

**Comment by tjdevries** [comment]
Ah, that's an interesting proposal.

I guess the reasoning would be that you'd like to be able to do something like click on the `modfile` identifier, and go to the import string itself at the top of the file? Instead of going directly to the module?

**Comment by JamyDev** [comment]
I was actually trying to mimick gopls' semantic tokens, which mark the actual variable as a definition. That also lets us do neat stuff like checking for unreferenced variables in a package without needing to actually parse the language.

**Comment by jupblb** [comment]
Having
```
  import (
   "golang.org/x/mod/modfile"
//  ^^^^^^^^^^^^^^^^^^^^^^^^ reference scip-go gomod golang.org/x/mod/modfile 0.0.1 golang.org/x/mod/modfile/
//                   ^^^^^^^ definition local 0
  )
```
is technically correct and follows `gopls` convention. Such a change may be breaking for `scip-go` consumers.

**Comment by jupblb** [comment]
I've experimented a bit with this idea at https://github.com/sourcegraph/scip-go/pull/197. It's a simple change but with puzzling tradeoffs. A few thoughts:
1. When going to the definition of a package prefix (e.g. `json` in `json.Marshal`) the current behavior is to move to some file with `package json` (exactly one file has `definition` entry for the `package`, the other have `reference`). We use some heuristic for making the decision which one contains `package` definition: https://github.com/sourcegraph/scip-go/blob/bbabeb50a5e931fe2c1fd2884e0547c0aeb5ad77/internal/index/package_ident.go#L13-L84 This is not ideal. An example of alternative behavior would be to make SCIP support having a relative directory as symbol definition. Such a change feels out of scope, for now.
2. Having local `definition` overlapping with `reference` may require complicated changes on SCIP index consumer end. It'd make it necessary to check for references for `local` symbol definitions.
3. Contrary to my previous message, the `gopls` conventions is to actually have something like the following:
    ```go
      import (
       "golang.org/x/mod/modfile"
    //  ^^^^^^^^^^^^^^^^^^^^^^^^ reference scip-go gomod golang.org/x/mod/modfile 0.0.1 golang.org/x/mod/modfile/
    //  ^^^^^^^^^^^^^^^^^^^^^^^^ definition local 0
      )
    ```
    Where the global reference has as many definitions as there are corresponding `package` statements.

It seems to me that going in the direction of `gopls` behavior is the right way to go but not because it's clearly the right solution. In case of `import` statement resolution we should have multiple definitions instead of an arbitrary one.

**Comment by JamyDev** [comment]
> We use some heuristic for making the decision which one contains package definition

Is there a hard requirement for a symbol being defined only once? I can see cases where multiple definition being an expected state (eg this package is defined across 5 files, so we should point out all 5 files in a gotodef (LSP [explicitly](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#textDocument_definition) allows multiple results)

> Having local definition overlapping with reference may require complicated changes on SCIP index consumer end

Would not returning a local even for aliased imports be the more canonically correct option? That would likely be the most compatible option.

> It seems to me that going in the direction of gopls behavior is the right way to go but not because it's clearly the right solution

I am thinking whether we want to reconsider the implications of the extra local indirection..

**Comment by jupblb** [comment]
> Is there a hard requirement for a symbol being defined only once?

No. In https://github.com/sourcegraph/scip-go/pull/197 I remove this limitation and mark all `package` occurrences as definitions.

> It seems to me that going in the direction of gopls behavior is the right way to go but not because it's clearly the right solution

I share the same opinion. :D
